### PR TITLE
Enhancement: distinct PayPal Donation live and sandbox connect flows

### PR DESCRIPTION
**Pull request by ravinderk**
## Description

In the pull request, I implement two connect buttons to connect the Paypal account in sandbox and live mode.

## Affects

This pull request affects PayPal donations onboarding feature.

## Visuals

image

## Testing Instructions

- [ ] You should be able to onboard with PayPal live mode.
- [ ] You should be able to connect with PayPal sandbox mode
- [ ] Existing connect PayPal account (live or sandbox) should display correctly on the setting page. 
- [ ] Donors should be able to donate with PayPal donations.
- [ ] Renewal should generate correctly for PayPal donation subscription in GiveWP test and live mode.
- [ ] Refresh cron job should exist for connected account type ( `give_paypal_commerce_refresh_live_token` or `give_paypal_commerce_refresh_sandbox_token`)
- [ ] Migration test
   - [ ] A refresh token cron job should exist for live  (`give_paypal_commerce_refresh_live_token`) or sandbox connection `give_paypal_commerce_refresh_sandbox_token` if PayPal connected
   - [ ] Legacy cron job should not exist `give_paypal_commerce_refresh_token`
- [ ] When the Paypal account is disconnected, related cron job, account details, and webhook details should be deleted from the database. Option name: `give_paypal_commerce_{$mode}_webhook_config`, `"give_paypal_commerce_{$mode}_account"`. Cron job: `give_paypal_commerce_refresh_${mode}_token`
## Pre-review Checklist

-   [ ] Acceptance criteria satisfied and marked in related issue
-   [x] Relevant `@unreleased` tags included in DocBlocks
-   [ ] Includes unit tests
-   [ ] Reviewed by the designer (if follows a design)
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1204754049884883
  - https://app.asana.com/0/0/1204621958568562

### CODE REVIEW

**Comment by JasonTheAdams** [comment]
@ravinderk Looking nice! I was able to connect with a sandbox account just fine, but I got the following when connecting my business account:

image

Are the capabilities still too strict in this PR? I'm guessing it is since https://github.com/impress-org/givewp/pull/6778 is still open. I've never connected my account to something like this before, so it's possible it's just not set up completely. Wouldn't it provide information on what's missing if that was the case? This error message is very strange.

I'll see if I can borrow someone's credentials to test it further.

File: src/PaymentGateways/PayPalCommerce/AdminSettingFields.php [hunk 1]
```diff
@@ -100,7 +95,7 @@ public function accountCountryField()
                 printf(
                     '%1$s<div class="give-field-description">%2$s</div>',
                     $settingHtml,
-                    esc_html__('The country your site operates from.', 'give')
+                    esc_html__('The country your site operates from. Select correct country before proceeding to connect', 'give')
```
Comment (JasonTheAdams) [inline src/PaymentGateways/PayPalCommerce/AdminSettingFields.php:98]:
```suggestion
                    esc_html__('The country the PayPal account is based from. Make sure to select before connecting.', 'give')
```

File: src/PaymentGateways/PayPalCommerce/AdminSettingFields.php [hunk 2]
```diff
@@ -117,99 +112,27 @@ public function payPalCommerceAccountManagerField()
...
+        $paypalLiveSetting->isRecurringAddonActive = $isRecurringAddonActive;
+        echo $this->getPayPalConnectionSettingView($paypalLiveSetting);
+
+        // Show sandbox PayPal connect button.
+        $paypalSandboxSetting = new \stdClass();
+        $paypalSandboxSetting->label = esc_html__('PayPal Sandbox Connection', 'give');
+        $paypalSandboxSetting->mode = 'sandbox';
+        $paypalSandboxSetting->connectButtonLabel = esc_html__('Connect with PayPal Sandbox', 'give');
+        $paypalSandboxSetting->description = esc_html__(
+            'PayPal sandbox is currently NOT connected. This is a separate sandbox PayPal account used for testing. Live PayPal accounts will not work.',
```
Comment (JasonTheAdams) [inline src/PaymentGateways/PayPalCommerce/AdminSettingFields.php:131]: In addition to wording, please have "PayPal Sandbox account" link to https://www.sandbox.paypal.com
```suggestion
            'PayPal sandbox is currently NOT connected. This is a separate PayPal Sandbox account, used for testing. Live PayPal accounts will not work.',
```

File: src/PaymentGateways/PayPalCommerce/ScriptLoader.php [hunk 3]
```diff
@@ -208,15 +203,13 @@ public function loadPublicAssets()
     /**
      * Get PayPal partner js url.
      *
+     * @unreleased sandbox PayPal partner js loads slow. So we are using live url for now.
```
Comment (JasonTheAdams) [inline src/PaymentGateways/PayPalCommerce/ScriptLoader.php:206]: Is it in any way a problem that we're not using the sandbox partner js?

**Review by JasonTheAdams — CHANGES_REQUESTED** [review]
Good work, @ravinderk! As mentioned, I wasn't able to connect a live account, but the UI is a lot less confusing.

File: src/PaymentGateways/PayPalCommerce/ScriptLoader.php (hunk 3 above)
Comment (ravinderk) [reply src/PaymentGateways/PayPalCommerce/ScriptLoader.php:206]: @JasonTheAdams It is not a problem. Paypal loads the same SDK in the sandbox and the live mode. Js SDK loads slowly in sandbox mode, so I used only the live mode JS. We can switch it manually if we need debugging.

**Review by JasonTheAdams — APPROVED** [review]
I'm not able to connect, but that very well could be my PayPal account. So let's get this off to QA and ask them to test both modes.

**Comment by ravinderk** [comment]
> I'm not able to connect, but that very well could be my PayPal account. So let's get this off to QA and ask them to test both modes.

@JasonTheAdams Can you check once? You should be able to connect the live PayPal account regardless of the GiveWP test mode status.

image


-------------------------

**Question:**
I findout that the token will be refreshed only for either sandbox or live mode detection based on the GiveWP test mode: https://github.com/impress-org/givewp/blob/5229306909ea2c7b2e33113929fc95b6f0d01c63/src/PaymentGateways/PayPalCommerce/RefreshToken.php#L71

We should have two cron jobs which independently refresh tokens for live and sandbox PayPal accounts (if connected).

Let me know if you have questions.

**Comment by Benunc** [comment]
@ravinderk see the comment on Asana: 

> If the site is in GiveWP test mode, the modal goes through the whole thing and then does what I describe in my last post ^^^^ (says it works, but then the page refreshes and it's still not connected)
> 
> But if I toggle the site into live mode, it works correctly. 
> 
> So the left-over check when the page refreshes to see if the site is in test mode is what's preventing it from saving correctly. I just toggled it to disable test mode and connected with no issue.

**Comment by ravinderk** [comment]
> > I'm not able to connect, but that very well could be my PayPal account. So let's get this off to QA and ask them to test both modes.
> 
> @JasonTheAdams Can you check once? You should be able to connect the live PayPal account regardless of the GiveWP test mode status.
> 
> image
> **Question:** I findout that the token will be refreshed only for either sandbox or live mode detection based on the GiveWP test mode:
> 
> https://github.com/impress-org/givewp/blob/5229306909ea2c7b2e33113929fc95b6f0d01c63/src/PaymentGateways/PayPalCommerce/RefreshToken.php#L71
> 
> We should have two cron jobs which independently refresh tokens for live and sandbox PayPal accounts (if connected).
> 
> Let me know if you have questions.



> @ravinderk see the comment on Asana:
> 
> > If the site is in GiveWP test mode, the modal goes through the whole thing and then does what I describe in my last post ^^^^ (says it works, but then the page refreshes and it's still not connected)
> > But if I toggle the site into live mode, it works correctly.
> > So the left-over check when the page refreshes to see if the site is in test mode is what's preventing it from saving correctly. I just toggled it to disable test mode and connected with no issue.

@Benunc I fixed the issue (https://github.com/impress-org/givewp/pull/6795#issuecomment-1592403705). I findout another issue with the access token refresh cron job. I will resolve it in a same pull request.

**Comment by angelablake** [comment]
Sorry, I didn't notice that, just saw that it passed QA in my backlog for releases. I'll only push the hot fix to master for release and not develop until Jason has had a chance to look it over.

File: src/PaymentGateways/PayPalCommerce/onBoardingRedirectHandler.php [hunk 4]
```diff
@@ -68,6 +68,24 @@ public function __construct(
...
+    /**
+     * This function set mode from request.
+     *
+     * @unreleased
+     * @return void
+     */
+    private function setModeFromRequest()
+    {
+        if (isset($_GET['mode']) && in_array($_GET['mode'], ['live', 'sandbox'])) {
+            $mode = sanitize_text_field(wp_unslash($_GET['mode']));
```
[the remark this answers is not in the document]
Comment (ravinderk) [reply src/PaymentGateways/PayPalCommerce/onBoardingRedirectHandler.php:84]: fixed: https://github.com/impress-org/givewp/commit/adb83329465575fa9f7b750f3132fd13c2413dab

**Pull request merged**